Transform graph redux

cbertinato opened the pull request:
This PR completely rewrites the transform infrastructure.  There is a lot more to be done, but in the interest of beginning to integrate the processing back end with the front end.  See the additions in the examples directory for examples of how initialize and interact with the transform graphs.

cbertinato committed:
Added graph and transform internals

cbertinato committed:
Changes to Graph class

cbertinato committed:
Added TransformGraph class

cbertinato committed:
Added tests and converted operations

cbertinato committed:
Refactoring in TransformGraph

cbertinato committed:
Refactored some in eotvos

cbertinato committed:
Added tests for TranformGraph subclassing

cbertinato committed:
Added sample period parameter for eotvos

cbertinato committed:
Added test for eotvos and fixed some bugs

cbertinato committed:
Added test for free air correction and fixed test for eotvos

cbertinato committed:
Added latitude correction

cbertinato committed:
Added test for use of partial in graph

cbertinato committed:
Added test for standard graph

cbertinato committed:
Added run method to TransformGraph

cbertinato committed:
TODO in interp_nans

cbertinato committed:
Added AirbornePost graph and removed standard gravity graph

cbertinato committed:
Change interpolate to use DataFrame method

cbertinato committed:
Removed old flowchart tests

cbertinato committed:
Added name to output series of lp_filter

cbertinato committed:
Some efficiency improvements

- More efficient method of generting the DatetimeIndex
- GPS sync status bit does not accurately indicate when the time is synced, so used another check for now

cbertinato committed:
Added verbosity flag to TransformGraph

cbertinato committed:
Fixed up some docstrings

cbertinato committed:
Added functionality for syncing gravity with gps acceleration

cbertinato committed:
Added option to align_frames for returning right, left, or both

cbertinato committed:
Fixed test to account for interpolation changes

cbertinato committed:
Fix tests

cbertinato committed:
Comments and TODOs

cbertinato committed:
Moved some functions out of processing graph

cbertinato committed:
Removed extraneous variables from AirbornePost

cbertinato committed:
Removed concatenation from AirbornePost graph

cbertinato committed:
Enabled specification of differentiator for eotvos

Commit:
ENH: Added process_script.py to examples directory for batch processing of airborne data.
ENH: Added plots.py file to lib directory.  Added shapely, cartopy, and geos to requirements.

Commit:
Temporarily removed time sync from airborne graph

Commit:
Removed comments

Commit:
Added second processing example

Commit:
Updated requirements

Also:
- Modified TravisCI configuration to install Cython in preparation for addition of the Cartopy package

bradyzp commented on tests/test_transform.py:
```diff
@@ -0,0 +1,223 @@
+# coding: utf-8
+import pytest
+from nose.tools import assert_almost_equals
```
Is it possible to replace this import/requirement with they pytest.approx or np.allclose functions?

cbertinato replied:
Done. Good call on this one.

bradyzp commented on requirements.txt:
```diff
@@ -1,13 +1,29 @@
+idna==2.6
 matplotlib==2.0.2
+more-itertools==4.2.0
+nose==1.3.7
```
See note in tests/test_transform (possible to remove this requirement and use pytest.approx or numpyp.allclose functions instead?)

bradyzp commented on dgp/lib/etc.py:
```diff
@@ -8,7 +8,7 @@
 
 def align_frames(frame1, frame2, align_to='left', interp_method='time',
-                 interp_only=[], fill={}):
+                 interp_only=[], fill={}, item='both'):
```
Possibly change the interp_only and fill defaults to None, then bind them to list/dict in the function body - there's a danger of unexpected behavior of the parameters are ever mutated.

cbertinato replied:
Good catch. Done.

bradyzp commented on tests/test_transform.py:
```diff
@@ -0,0 +1,223 @@
+                           }
+        g = TransformGraph(graph=transform_graph)
+        res = g.execute()
+        np.testing.assert_array_almost_equal(expected, res['fac'], decimal=8)
```
I'm experiencing a test failure here on my system, the actual result value is identical to the expected, but with the wrong sign (positive instead of negative).

cbertinato replied:
That's odd. Let's talk online tomorrow. I get a passing test in my environment and I'll try to think of why that might be in the meantime.

bradyzp commented on dgp/lib/transform/transform.py:
```diff
@@ -0,0 +1,179 @@
+import inspect
+from functools import wraps
+
+from dgp.lib.etc import gen_uuid, dedup_dict
```
dedup_dict seems to be missing from this branch, it looks to have been introduced in feature/data-transform commit# 0623cca

cbertinato replied:
Hm. Must have accidentally deleted it in the rebase. Done.

bradyzp commented on dgp/lib/transform/derivatives.py:
```diff
@@ -17,44 +17,17 @@ def centraldifference(data_in, n=1, order=2, dt=0.1):
+    """ 10th order Taylor series FIR differentiator """
+    coeff = np.array([1 / 1260, -5 / 504, 5 / 84, -5 / 21, 5 / 6, 0, -5 / 6, 5 / 21, -5 / 84, 5 / 504, -1 / 1260])
+    x = data_in
+    for _ in range(1, n + 1):
```
Should be able to remove 'y' variable and replace with
```
x = convolve(x, coeff, mode='same')
```
Fixes linter warning about y 'might be referenced before assignment'

cbertinato replied:
Done. In fact, made it even a bit cleaner. I don't frankly know why I did it that way.

bradyzp reviewed (changes requested):
Looks great, the only significant issue I see is the missing dedup_dict method in lib/etc.py, everything else is minor and can be touched up later.

Also is install_proj4.sh needed for anything - it's just an empty file currently.

Commit:
Added back dedup_dict function

Commit:
Removed install_proj4.sh

Commit:
Changed default values for interp_only and fill

The defaults values for interp_only and fill arguments are now None and then bound to an empty list and dictionary, respectively, in the function.

Commit:
Swapped assert_almost_equals for pytest.approx

Commit:
Minor change in taylor_fir for pylint warning
